### PR TITLE
AWS S3 Provider

### PR DESCRIPTION
**Pull request by andymac4182**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/SixLabors/ImageSharp.Web/pulls) open
- [x] I have verified that I am following matches the existing coding patterns and practice as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [ ] I have provided test coverage for my change (where applicable)

### Description

Hi,

We are internally using our own provider based on some of our own libraries. I am looking to port this to the latest version of `ImageSharp.Web` and remove our custom libraries. Is this something you would be interested in supporting in the library?

For testing internally we use [Minio](https://min.io) which I think would fit with how you are testing Azure Blob Storage.

Cheers,
Andrew

### CODE REVIEW

**Comment by JimBobSquarePants** [comment]
Thanks @andymac4182 I'll have a proper look asap. Will do a cursory review just now.

**Comment by andymac4182** [comment]
Sorry for locking up the build with the mistake in the `appveyor.yml` I am pretty sure I have fixed it up.

File: src/ImageSharp.Web.Providers.AWS/Providers/AWSS3StorageImageProvider.cs [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+            string imageId = Path.GetFileNameWithoutExtension(displayUrl);
+            this.logger.LogDebug("Image id is {ImageId}", imageId);
+
+            string imagePath = $"{imageId}.png";
+
+            bool imageExists = await this.KeyExists(this.storageOptions.BucketName, imagePath);
+
+            if (!imageExists)
+            {
+                this.logger.LogDebug("No image found for {ImageId}", imageId);
```
Comment (JimBobSquarePants) [inline src/ImageSharp.Web.Providers.AWS/Providers/AWSS3StorageImageProvider.cs]: Don't need these logger statements here. We log in the middleware.

File: src/ImageSharp.Web.Providers.AWS/Providers/AWSS3StorageImageProvider.cs [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+        /// <inheritdoc />
+        public async Task<IImageResolver> GetAsync(HttpContext context)
+        {
+            PathString displayUrl = context.Request.Path;
+            this.logger.LogDebug("Getting image for {ImageUri}", displayUrl);
+            string imageId = Path.GetFileNameWithoutExtension(displayUrl);
+            this.logger.LogDebug("Image id is {ImageId}", imageId);
+
+            string imagePath = $"{imageId}.png";
```
Comment (JimBobSquarePants) [inline src/ImageSharp.Web.Providers.AWS/Providers/AWSS3StorageImageProvider.cs]: Png only?

File: src/ImageSharp.Web.Providers.AWS/Providers/AWSS3StorageImageProvider.cs [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+            try
+            {
+                var request = new GetObjectMetadataRequest
+                {
+                    BucketName = bucketName,
+                    Key = key
+                };
+
+                ((Amazon.Runtime.Internal.IAmazonWebServiceRequest)request)
```
Comment (JimBobSquarePants) [inline src/ImageSharp.Web.Providers.AWS/Providers/AWSS3StorageImageProvider.cs]: I don't this is correct. The internal namespace is a big red flag.

File: src/ImageSharp.Web.Providers.AWS/Providers/AWSS3StorageImageProviderOptions.cs [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+    /// <summary>
+    /// Configuration options for the <see cref="AWSS3StorageImageProvider"/> provider.
+    /// </summary>
+    public class AWSS3StorageImageProviderOptions
+    {
+        /// <summary>
+        /// Gets or sets the bucket name.
+        /// Must conform to AWS S3 bucket naming guidelines.
+        /// </summary>
+        public string BucketName { get; set; }
```
Comment (JimBobSquarePants) [inline src/ImageSharp.Web.Providers.AWS/Providers/AWSS3StorageImageProviderOptions.cs]: Credentials?

File: src/ImageSharp.Web.Providers.AWS/Providers/AWSS3StorageImageProvider.cs (hunk 2 above)
Comment (andymac4182) [reply src/ImageSharp.Web.Providers.AWS/Providers/AWSS3StorageImageProvider.cs]: Sorry this was a left over from our internal code. I will remove the png only code.

File: src/ImageSharp.Web.Providers.AWS/Providers/AWSS3StorageImageProvider.cs (hunk 1 above)
Comment (andymac4182) [reply src/ImageSharp.Web.Providers.AWS/Providers/AWSS3StorageImageProvider.cs]: Too easy will clean these up.

File: src/ImageSharp.Web.Providers.AWS/Providers/AWSS3StorageImageProvider.cs (hunk 3 above)
Comment (andymac4182) [reply src/ImageSharp.Web.Providers.AWS/Providers/AWSS3StorageImageProvider.cs]: This function came from their library. I will check but I think I can remove this section.

File: src/ImageSharp.Web.Providers.AWS/Providers/AWSS3StorageImageProviderOptions.cs (hunk 4 above)
Comment (andymac4182) [reply src/ImageSharp.Web.Providers.AWS/Providers/AWSS3StorageImageProviderOptions.cs]: I was going to do some reading. AWS has multiple ways of initialising the client. Their recommendation is normally to use their extension methods. I will see how I go. We register a single S3 client outside of all of our code and I was thinking of applying something similar but I will confirm through their doco.

**Comment by andymac4182** [comment]
I have seen how your testing the Azure provider. I will add some tests this week for the AWS one in a similar manner.

**Comment by JimBobSquarePants** [comment]
@andymac4182 Any possibility we can get this up and running again? It would be nice to ship it with V1.

It's a brave new world using Github Actions for CI so I can help get that working.

**Comment by breaker05** [comment]
Any update on this being implemented?

**Comment by phinett** [comment]
> Any update on this being implemented?

Also keen to get this working.  I have tried to add this code to my own project, it compiles and runs, but i can't get it to hit any breakpoints in the provider (apart from the constructor on initial load).

I'm not totally sure howi should access the image though, my bucket name is "my-s3" for instance, and im accessing http://localhost:5000/my-s3/myimage.jpg

Am i missing something?

**Comment by JimBobSquarePants** [comment]
I really do not appreciate these "Any updates" kind of questions. They paint a picture of consumption without contribution which is not healthy for OSS.

@andymac4182 is doing this work in his own time and is under no obligation to ever complete it. (_Please note, neither am I_). I suggest that if there is any urgency on your part you should invest the time personally to implement a working solution.

**Comment by phinett** [comment]
I'm not asking for an update on when its done.

I am actually trying to have a go at this myself, was just after a point in the right direction of how this pieces together more than anything.

Do i need to implement cache providers, file storage provider etc? Or should it simply work by registering an IImageProvider?

**Comment by breaker05** [comment]
@JimBobSquarePants, my only intention about asking for an update was because this PR on your project is from 16 months ago and is a working solution so was not sure why it was not considered for your v1 release.

You immediately thumbs downed my response of me asking a very simple question if this work by @andymac4182 would be merged and implemented. I have no context as to why you did this since you left no comment. When I saw this earlier I took it as you had no intentions of using the work from @andymac4182. I instead wrote my own implementation and have it working successfully and happy to do my own PR, even though quite redundant to the work already here.

Your library is wonderful and I am quite thankful for it with the ability to extend it with other providers. I only ask you to not rush to judgment that everyone is only out to consume. I am happy to contribute to this project as well.

**Comment by JimBobSquarePants** [comment]
@phinett @breaker05 

Thanks both for your interest and also for expressing the interest in contributing. 

Please both be aware though that you initial comments do not, (_from a maintainers perspective_), match the intent you have both since expressed and that there is no way, (_from a maintainers perspective_), to interpret the intent in a positive manner. 

We all have something to learn from the exchange.

That aside...

@breaker05 If you have a working solution that you are able to share that would be wonderful! 

In order to get something that is mergeable though we need a simple way to unit test your code (locally and in GitHub Actions) without having to set up an AWS account. That was the main reason for dropping the initial work from the V1 build. 

Azure makes this very simple and need to be able to do the same for AWS.

If that is the case then I will happily work with you to integrate the solution into the repository here.

**Pull request merged**